File: backend/src/main/java/com/level30/api/domain/engine/RiskEngine.java
package com.level30.api.domain.engine;

import com.level30.api.domain.model.Challenge;
import com.level30.api.domain.model.RiskAssessment;
import com.level30.api.domain.model.RiskLevel;
import com.level30.api.domain.model.SuggestedAction;
import java.time.Duration;
import java.time.Instant;
import java.util.Set;
import org.springframework.stereotype.Component;

/**
 * Motor de risco de abandono.
 *
 * <p><b>ESPELHO EXATO</b> de:
 * <ul>
 *   <li>{@code lib/domain/engine/risk_engine.dart} (Flutter)</li>
 *   <li>{@code server/src/risk.ts} (Cloudflare Worker)</li>
 * </ul>
 * Mesmos pesos, mesmos thresholds, mesma heurística de dias-sem-atividade.
 * Ao alterar um, alterar os outros dois — e a suíte {@code RiskEngineTest}
 * é o contrato compartilhado (portada de {@code test/risk_engine_test.dart}).
 */
@Component
public class RiskEngine {

    private static final Set<Integer> MILESTONES = Set.of(7, 14, 21, 30);

    public RiskAssessment assess(Challenge challenge) {
        double score = calculateScore(challenge);
        RiskLevel level = scoreToLevel(score);
        SuggestedAction action = determineAction(challenge, level);
        return new RiskAssessment(
                challenge.getId(),
                score,
                level,
                action,
                Instant.now()
        );
    }

    public double calculateScore(Challenge c) {
        double score = 0.0;

        // Fator 1: dias sem atividade (40%)
        int daysSince = daysSinceLastActivity(c);
        score += switch (daysSince) {
            case 0 -> 0.0;
            case 1 -> 0.1;
            case 2 -> 0.25;
            default -> 0.4;
        };

        // Fator 2: taxa de progresso vs esperada (30%)
        if (c.getTotalDays() > 0) {
            double rate = (double) c.getCurrentDay() / c.getTotalDays();
            score += (1.0 - rate) * 0.3;
        }

        // Fator 3: streak (30%)
        double streakFactor = c.getStreak() == 0
                ? 0.3
                : Math.max(0.0, 0.3 - c.getStreak() * 0.03);
        score += streakFactor;

        return clamp01(score);
    }

    public RiskLevel scoreToLevel(double score) {
        if (score < 0.25) {
            return RiskLevel.LOW;
        }
        if (score < 0.50) {
            return RiskLevel.MEDIUM;
        }
        if (score < 0.75) {
            return RiskLevel.HIGH;
        }
        return RiskLevel.CRITICAL;
    }

    public SuggestedAction determineAction(Challenge c, RiskLevel level) {
        if (MILESTONES.contains(c.getCurrentDay())) {
            return SuggestedAction.CELEBRATE_MILESTONE;
        }
        return switch (level) {
            case LOW -> SuggestedAction.NONE;
            case MEDIUM -> SuggestedAction.SEND_REMINDER;
            case HIGH -> SuggestedAction.SEND_MOTIVATION;
            case CRITICAL -> SuggestedAction.SUGGEST_REPLAN;
        };
    }

    private int daysSinceLastActivity(Challenge c) {
        if (c.getLastActivityAt() == null) {
            return c.getStreak() == 0 ? 2 : 0;
        }
        return (int) Duration.between(c.getLastActivityAt(), Instant.now()).toDays();
    }

    private double clamp01(double v) {
        return Math.min(1.0, Math.max(0.0, v));
    }
}

File: lib/domain/engine/risk_engine.dart
import '../../data/model/challenge.dart';
import '../../data/model/risk_assessment.dart';

class RiskEngine {
  RiskAssessment assess(Challenge challenge) {
    final score = _calculateScore(challenge);
    final level = _scoreToLevel(score);
    final action = _determineAction(challenge, level);
    return RiskAssessment(
      challengeId: challenge.id,
      riskScore: score,
      riskLevel: level,
      suggestedAction: action,
      calculatedAt: DateTime.now(),
    );
  }

  double _calculateScore(Challenge c) {
    double score = 0.0;

    // Fator 1: dias sem atividade (40%)
    final daysSince = _daysSinceLastActivity(c);
    score += switch (daysSince) {
      0 => 0.0,
      1 => 0.1,
      2 => 0.25,
      _ => 0.4,
    };

    // Fator 2: taxa de progresso vs esperada (30%)
    if (c.totalDays > 0) {
      final rate = c.currentDay / c.totalDays;
      score += (1.0 - rate) * 0.3;
    }

    // Fator 3: streak (30%)
    final streakFactor =
        c.streak == 0 ? 0.3 : (0.3 - (c.streak * 0.03)).clamp(0.0, 0.3);
    score += streakFactor;

    return score.clamp(0.0, 1.0);
  }

  RiskLevel _scoreToLevel(double s) => switch (s) {
        < 0.25 => RiskLevel.low,
        < 0.50 => RiskLevel.medium,
        < 0.75 => RiskLevel.high,
        _ => RiskLevel.critical,
      };

  SuggestedAction _determineAction(Challenge c, RiskLevel level) {
    if (const [7, 14, 21, 30].contains(c.currentDay)) {
      return SuggestedAction.celebrateMilestone;
    }
    return switch (level) {
      RiskLevel.low => SuggestedAction.none,
      RiskLevel.medium => SuggestedAction.sendReminder,
      RiskLevel.high => SuggestedAction.sendMotivation,
      RiskLevel.critical => SuggestedAction.suggestReplan,
    };
  }

  int _daysSinceLastActivity(Challenge c) {
    if (c.lastActivityAt == null) return c.streak == 0 ? 2 : 0;
    return DateTime.now().difference(c.lastActivityAt!).inDays;
  }
}
